add: Add streamlined module creation feature

MAVRICK-1 opened the pull request:
# Fix: Add streamlined module creation feature

## Overview
Implements a direct "Create Module" workflow in the trigger editor, eliminating the clunky export-then-import process described in issue #566.

## Changes Made

### Files Modified:
- **dlgTriggerEditor.h** - Added new action and slot declarations
- **dlgTriggerEditor.cpp** - Implemented "Create Module" button and functionality  
- **dlgPackageExporter.h** - Added preselection and module creation mode methods
- **dlgPackageExporter.cpp** - Enhanced to support module creation workflow

### Key Features:
1. **New "Create Module" button** in trigger editor toolbar (📦 icon)
2. **Context-aware preselection** - automatically selects current item
3. **Module creation mode** - configures package exporter for module workflow
4. **Auto-installation** - modules are installed immediately after creation
5. **Streamlined UX** - single-click process from selection to installed module

## Implementation Details

### dlgTriggerEditor Changes:
- Added `mpCreateModuleAction` toolbar button with package-exporter icon
- Implemented `slot_createModule()` that opens package exporter in module mode
- Pre-selects current item based on editor view type (Trigger/Alias/Script/etc.)

### dlgPackageExporter Enhancements:
- Added `setModuleCreationMode()` to configure UI for module creation
- Added preselect methods for all item types (Trigger, Timer, Alias, Script, Action, Key)
- Modified export completion to auto-install as module when in module creation mode
- Updated window title and prompts for module creation workflow


## Related Issues
Closes #566
/claim #566

## Screenshots
*Note: Screenshots can be added showing the new Create Module button in the trigger editor toolbar and the enhanced package exporter dialog in module creation mode.*
<img width="1920" height="1200" alt="Screenshot from 2025-08-16 20-05-59" src="https://github.com/user-attachments/assets/560a3e29-1a91-4a58-a930-94474701d861" />

https://github.com/user-attachments/assets/dbaa7019-c853-4d33-86be-5c39dd4c1fa4

<img width="1920" height="1200" alt="image" src="https://github.com/user-attachments/assets/631e202c-7d05-4ba9-8c03-7d33e90d7dfd" />

MAVRICK-1 committed:
feat: Add streamlined module creation feature

  - Add 'Create Module' button to trigger editor toolbar
  - Implement one-click module creation workflow
  - Auto-preselect current item in package exporter
  - Auto-install modules after creation
  - Eliminate clunky export-then-import process

  Fixes #566

MAVRICK-1 commented:
cc @vadi2 is the approach correct ?

vadi2 commented:
@MAVRICK-1 it's a good approach, I suggest to refine it so that when you're creating a module the dialog isn't talking about exporting it (as that's not what you're doing).

MAVRICK-1 commented:
> @MAVRICK-1 it's a good approach, I suggest to refine it so that when you're creating a module the dialog isn't talking about exporting it (as that's not what you're doing).


@vadi2 
Thanks for the feedback! I want to make sure I understand correctly - when you mention "refining it so that when creating a module the dialog isn't talking about exporting it," could you help me understand what  you're referring to ?

vadi2 commented:
This dialog shows the module creation dialog, but at the same time it says things not related to module creation - 'update installed package', 'select export location', 'export'. This will confuse players :slightly_smiling_face: 

<img width="1920" height="1200" alt="image" src="https://github.com/user-attachments/assets/e06c794e-82e7-472d-8592-307ea20715b1" />

MAVRICK-1 committed:
feat: Enhance module creation UI with updated labels and descriptions

MAVRICK-1 commented:
<img width="1920" height="1200" alt="Screenshot from 2025-08-17 20-11-20" src="https://github.com/user-attachments/assets/2e907cb7-03e4-4977-85c9-921eed6496a4" />
@vadi2 done

MAVRICK-1 committed:
Merge branch 'development' into feature/streamlined-module-creation

vadi2 commented on src/dlgPackageExporter.cpp:
```diff
@@ -165,10 +165,74 @@ void dlgPackageExporter::setModuleCreationMode(bool isModule)
+        while (iter != modules.constEnd()) {
+            ui->DependencyList->addItem(iter.key());
+            ++iter;
+        }
```
I would prefer a modern C++ for loop here

vadi2 commented on src/dlgPackageExporter.cpp:
```diff
@@ -159,6 +159,161 @@ dlgPackageExporter::~dlgPackageExporter()
+        ui->pushButton_openInfos->setText(tr("Add module description, icon, and assets (optional)"));
+        
+        // Update the export location widget to be more descriptive
+        ui->lineEdit_filePath->setPlaceholderText(tr("Module will be saved here"));
```
```suggestion
        ui->lineEdit_filePath->setPlaceholderText(tr("Module location"));
```

vadi2 commented on src/dlgPackageExporter.cpp:
```diff
@@ -159,6 +159,161 @@ dlgPackageExporter::~dlgPackageExporter()
+        ui->lineEdit_filePath->setPlaceholderText(tr("Module will be saved here"));
+        
+        // Update info label if visible
+        ui->infoLabel->setText(tr(""));
```
```suggestion
        ui->infoLabel->setText(QString());
```

vadi2 committed:
Merge branch 'development' into feature/streamlined-module-creation

vadi2 commented on src/dlgPackageExporter.cpp:
```diff
@@ -159,6 +159,161 @@ dlgPackageExporter::~dlgPackageExporter()
+
+void dlgPackageExporter::preselectTimer(QTreeWidgetItem* item)
+{
+    if (!item) return;
```
Format the file with clang-tidy please, we have a .clang-tidy file [here](https://github.com/Mudlet/Mudlet/blob/development/.clang-tidy). This and other places should be:

```suggestion
    if (!item) {
      return;
    }
```

vadi2 reviewed (commented):
This seems to hang exporting for me -

https://github.com/user-attachments/assets/ae066490-b40f-4c27-9c6f-7e2deeebd4a7

Also when I click `Create module` then 300+ items are already selected - why is that?

MAVRICK-1 commented:
okay on it

MAVRICK-1 committed:
Update src/dlgPackageExporter.cpp

Co-authored-by: Vadim Peretokin <vperetokin@hey.com>

MAVRICK-1 committed:
Update src/dlgPackageExporter.cpp

Co-authored-by: Vadim Peretokin <vperetokin@hey.com>

MAVRICK-1 committed:
Fix review comments for Create Module feature

- Replace while loop with modern C++ range-based for loop in dependency list
- Fix QString() usage instead of tr("") for empty strings
- Fix 300+ items auto-selection issue by disconnecting signals during UI setup
- Add proper QFutureWatcher cleanup to prevent hanging exports
- Improve error handling in async export operations

Addresses reviewer feedback on PR

MAVRICK-1 commented:
https://github.com/user-attachments/assets/364b5bdc-dabd-4d8b-ba8d-2462b11ca01a

@vadi2  i fixed it 


### Issue A: Unintended Mass Selection During Module Creation

**Problem Description:**
When users clicked the "Create Module" button, the system incorrectly auto-selected hundreds of items instead of selecting only the currently active item.

**Root Cause Analysis:**
The issue occurred when `ui->packageList->clear()` was called, which triggered the `currentIndexChanged` signal. This signal activated the `slot_packageChanged()` method with an empty package name parameter. The method then proceeded to select all items that had empty package names, resulting in the mass selection of unintended items.

**Resolution:**
Implemented a temporary signal disconnection during the UI setup phase in module creation mode to prevent the unintended signal cascade.

---

### Issue B: Module Export Process Hanging

**Problem Description:**
The module export functionality would become unresponsive and fail to complete, leaving users unable to export their modules successfully.

**Root Cause Analysis:**
The hanging behavior was caused by improper cleanup of `QFutureWatcher` objects. These objects were not being properly disposed of after use, leading to memory leaks and preventing asynchronous operations from completing correctly.

**Resolution:**
Added proper cleanup procedures by implementing `watcher->deleteLater()` calls to ensure `QFutureWatcher` objects are properly destroyed and memory is released, allowing export operations to complete successfully.

vadi2 commented:
Thanks, you can just show the fix in the future! I don't need to read paragraphs of LLM-generated text as well - it's cheap to generate but takes effort to read

vadi2 commented:
If you export a module multiple times, it will start to duplicate the contents. Here I exported it 4 times:

<img width="1105" height="530" alt="image" src="https://github.com/user-attachments/assets/74b1ac5c-f0dd-4aff-9180-989814dc55c4" />

MAVRICK-1 commented:
![Screenshot_2025-08-19-01-12-15-559_com anthropic claude](https://github.com/user-attachments/assets/4a22dec0-7c5a-4b34-9aaa-304cc908936e)
@vadi2 i just optimised the markdown using AI 😅, sorry will keep in mind the next time

MAVRICK-1 committed:
Fix review comments and improve module creation workflow

- Fix modern C++ for loop to use range-based iteration
- Update placeholder text to 'Module location'
- Change empty tr() to QString()
- Fix auto-selection issue preventing 300+ items from being selected
- Add smart duplicate detection to handle stale references
- Implement proper overwrite dialog with user choice
- Remove unwanted unpacking dialog for UI-created modules
- Add success dialog with auto-close functionality
- Prevent duplicate content creation in modules

Fixes all maintainer review comments and improves user experience.

MAVRICK-1 commented:
@vadi2 i have fix the duplicate issue and also add a new workflow

https://github.com/user-attachments/assets/8009c3c1-0e8a-4737-95f3-8fab038e3f73

MAVRICK-1 commented:
https://github.com/user-attachments/assets/4b6ff7be-ac2e-4729-af26-640f3682a18c

i have also added this if we select a  trigger/ item then click on create module it will automatically select it

vadi2 commented:
Nice I think this is looking good! I will test the code later. UI wise, pop up dialogs are discourages in modern UX design, do you think you could make the messages be embedded within the dialog itself? 

This is similar to how the package export works currently - it doesn't create a pop-up when it's successful, but has an embedded message instead.

MAVRICK-1 commented:
@vadi2 The message will be displayed at the bottom of the dialog box, which should not be good in terms of UI. Important messages like 'duplicate module' should be handled with pop-up dialog boxes. I was just following the workflow of VS Code. If you want me to change it, I can do it.

vadi2 commented:
A 'module has been created and installed successfully' is not that important, it's more of an information message.

Can you show me an example in VS Code where it does pop ups like this?

MAVRICK-1 commented:
https://github.com/user-attachments/assets/6adc9b6a-53d3-4285-a9ca-d07f4782ee22

 i am taking the example of delete because overwriting is not available there. You can also get the overwrite example while copying something if it already exists.

vadi2 commented:
I see what you mean - keeping it for overwrite would be OK, but we can embed "everything is OK" message.

MAVRICK-1 commented:
> I see what you mean - keeping it for overwrite would be OK, but we can embed "everything is OK" message.

just to clear here are you taking about Module created succesful message ? yes i can do that

vadi2 committed:
Merge branch 'development' into feature/streamlined-module-creation

MAVRICK-1 commented:
https://github.com/user-attachments/assets/873d21dc-24d6-4aa0-9514-5eccee01cf5e

@vadi2 will add auto close after 2 sec or the user will close it ?

vadi2 commented:
Think we can leave it to the user to close.

MAVRICK-1 committed:
Fix module persistence and add empty module validation

Core Issues Fixed:
- Module persistence: Fixed XMLimport.cpp to populate both mInstalledModules and mActiveModules during profile load
- Empty module validation: Added comprehensive validation in dlgPackageExporter.cpp with embedded error messages
- UX improvements: Replaced popup dialogs with embedded success/error messages

Testing Results:
✅ Modules now persist correctly across Mudlet restarts
✅ Empty module attempts show proper validation error
✅ Module Manager displays previously created modules
✅ Clean embedded messaging without intrusive popups

Note: Module file location (saving to build dir vs profile dir) needs investigation

Addresses core functionality issues for GitHub #566

MAVRICK-1 committed:
Fix module save location - final issue resolved

Problem: Modules were saving to build directory instead of profile directory
Solution: Modified dlgPackageExporter.cpp to use profile directory for modules

Changes:
- Modules now save to ~/.config/mudlet/profiles/ProfileName/
- Packages still save to user's chosen directory
- Added conditional logic based on mIsModuleCreationMode

Testing Results:
✅ Verified modules save to correct profile directory
✅ No more .mpackage files in build directory
✅ All module functionality working perfectly

ALL ISSUES NOW RESOLVED! 🎉
- ✅ Module persistence across restarts
- ✅ Empty module validation
- ✅ Embedded messaging (no popups)
- ✅ Correct save locations
- ✅ Module Manager displays all modules

The streamlined module creation feature is now complete and ready for production use.

Fixes GitHub #566 completely

MAVRICK-1 commented:
https://github.com/user-attachments/assets/3bcd1408-1c75-4bbf-8ed3-5f69bb05a8f7

cc @vadi2

vadi2 reviewed (approved):
Looks great to me, thanks. Will give others a chance to review too.

MAVRICK-1 committed:
Fix stale module reference cleanup for robust duplicate detection

Problem: Ghost module references caused 'already exists' errors for deleted modules
Solution: Enhanced stale reference detection to check both directory and file paths

This ensures old broken module references are properly cleaned up when users
try to recreate modules with the same names.

MAVRICK-1 committed:
Fix empty if statement body in Host.cpp

Removed empty if statement that was left after debug cleanup.
Now simply calls saveProfile() without error handling since
this is a fire-and-forget profile save operation.

MAVRICK-1 committed:
Merge branch 'development' into feature/streamlined-module-creation
